usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+        self._trigger()
+        return self
+
+
+class Timeout(Event[V]):
+    """
+    Automatically triggered Event that processes can wait for
+
+    This event does automatically :py:meth:`~.succeed` after ``delay``
+    has passed. It should not be triggered manually.
+
+    :param delay: time until the event is triggered
+    :param value: value of the event when triggered
+
+    .. hint::
+
+        **Migrating to μSim**
+
+        Time related notifications can be created using :py:data:`usim.time`.
+
+        **timeout**
+            Use ``await (time + delay)``.
+
+        **deadline**
+            Use ``await (time >= deadline)`` or ``await (time == deadline)``.
+    """
+    __slots__ = '_fixed_value', '_delay'
+
+    def __init__(self, env, delay, value: Optional[V] = None):
+        if delay < 0:
+            raise ValueError("'delay' must not be negative")
+        super().__init__(env)
+        self._delay = delay
+        self._fixed_value = value
+        env.schedule(self._trigger_timeout())
+
+    async def _trigger_timeout(self):
+        await (time + self._delay)
+        self.succeed(self._fixed_value)
+
+    def __repr__(self):
+        return f'<usim.py.{self.__class__.__name__} delay={self._delay}>'
+
+
+class Initialize:
+    """
+    Internal initializer for :py:class:`~.Process`
+
+    Since this event is internal and not used by 'usim.py',
+    it is not implemented. Instantiating this class
+    unconditionally raises an error.
+    """
+    def __init__(self, *args, **kwargs):
+        raise NotEmulatedError(self.__doc__)
+
+
+class InterruptQueue:
+    """
+    Internal helper to manage interrupts for :py:class:`~.Process`
+    """
+    __slots__ = '__usimpy_flag__', '_causes', 'ok', 'defused'
+
+    def __init__(self):
+        self.__usimpy_flag__ = Flag()
+        self._causes = []
+        self.ok = False
+        self.defused = True
+
+    def __bool__(self):
+        return bool(self._causes)
+
+    @property
+    def value(self):
+        """The next :py:exc:`Interrupt`; this is a destructive action"""
+        return Interrupt(self.pop())
+
+    def push(self, cause):
+        """Add a new interrupt with ``cause``"""
+        self._causes.append(cause)
+        if not self.__usimpy_flag__:
+            self.__usimpy_flag__._value = True
+            self.__usimpy_flag__.__trigger__()
+
+    def pop(self):
+        result = self._causes.pop(0)
+        if not self._causes:
+            self.__usimpy_flag__._value = False
+        return result
+
+
+class Process(Event[V]):
+    """
+    A concurrently running process which triggers as an event on completion
+
+    Processes wrap around generators, which in turn may ``yield`` events to
+    wait for them. This implements cooperative multitasking, allowing another
+    process to run while others wait. Each :py:class:`~.Process` acts as an
+    event which is automatically triggered when the underlying generator completes.
+
+    .. code:: python3
```

Comment:
I am afraid the examples don't speak for themselves. What do you want to show? E.g. what is the purpose of `clocks`? Is it only to show the difference to the `multi_clock` function below? Then I wouldn't get the point when reading until here.